augur/filter.py

```diff
@@ -619,272 +1076,419 @@ def register_arguments(parser):
             "ERROR: You need to provide a sequence index or sequences to filter on sequence-specific information.",
             file=sys.stderr)
-        return 1
+        return False
 
-    # Load inputs, starting with metadata.
-    try:
-        # Metadata are the source of truth for which sequences we want to keep
-        # in filtered output.
-        metadata, meta_columns = read_metadata(args.metadata, as_data_frame=True)
-        metadata_strains = set(metadata.index.values)
-    except ValueError as error:
-        print("ERROR: Problem reading in {}:".format(args.metadata))
-        print(error)
-        return 1
-
-    #Set flags if VCF
-    is_vcf = False
-    is_compressed = False
-    if args.sequences and any([args.sequences.lower().endswith(x) for x in ['.vcf', '.vcf.gz']]):
-        is_vcf = True
-        if args.sequences.lower().endswith('.gz'):
-            is_compressed = True
+    # Set flags if VCF
+    is_vcf = filename_is_vcf(args.sequences)
 
     ### Check users has vcftools. If they don't, a one-blank-line file is created which
     #   allows next step to run but error very badly.
     if is_vcf:
         from shutil import which
         if which("vcftools") is None:
             print("ERROR: 'vcftools' is not installed! This is required for VCF data. "
-                  "Please see the augur install instructions to install it.")
-            return 1
+                  "Please see the augur install instructions to install it.",
+                  file=sys.stderr)
+            return False
 
-    # Read in files
+    return True
 
-    # If VCF, open and get sequence names
-    if is_vcf:
-        vcf_sequences, _ = read_vcf(args.sequences)
-        sequence_strains = set(vcf_sequences)
-    elif (args.sequences and any(SEQUENCE_ONLY_FILTERS)) or args.sequence_index:
-        # If FASTA, try to load the sequence composition details and strain
-        # names to be filtered.
-        index_is_autogenerated = False
-        sequence_index_path = args.sequence_index
 
+def run(args):
+    '''
+    filter and subsample a set of sequences into an analysis set
+    '''
+    # Validate arguments before attempting any I/O.
+    if not validate_arguments(args):
+        return 1
+
+    # Determine whether the sequence index exists or whether should be
+    # generated. We need to generate an index if the input sequences are in a
+    # VCF, if sequence output has been requested (so we can filter strains by
+    # sequences that are present), or if any other sequence-based filters have
+    # been requested.
+    sequence_strains = None
+    sequence_index_path = args.sequence_index
+    build_sequence_index = False
+    is_vcf = filename_is_vcf(args.sequences)
+
+    if sequence_index_path is None and args.sequences:
+        sequence_filters_requested = any(getattr(args, arg) for arg in SEQUENCE_ONLY_FILTERS)
+        if is_vcf or args.output or sequence_filters_requested:
+            build_sequence_index = True
```

Comment:
I just ran into an issue while testing this code with the full ncov dataset, where I wanted to extract sequences and metadata from the full GISAID database for a given list of strains (i.e., I don't want any filters, I want to exclude-all and then force-include strains from a list).

This command should have done the job quickly:

```bash
augur filter \
  --sequences filtered.fasta.xz \
  --metadata metadata.tsv.gz \
  --exclude-all \
  --include nebraska_strains.txt \
  --output-metadata nebraska_metadata.tsv.gz \
  --output-sequences nebraska_filtered.fasta.gz
```

Instead, I got the following notice from augur:

```
Note: You did not provide a sequence index, so Augur will generate one. You can generate your own index ahead of time with `augur index` and pass it with `augur filter --sequence-index`.
```

I am torn here about the expected behavior for handling sequences. In the code above, I chose to build a sequence index when the user has requested any sequence output even when the user hasn't requested any sequence-based filters. My reasoning at the time was that the current API in v12.0.0 will filter out strains from metadata and strain outputs if those strains are missing from the sequence inputs. Because we stream metadata and strains to disk in the first pass before looping through sequence data, we don't know if strains are missing from the sequence data unless we first build a sequence index.

A clearer way to express the general desired interface might be:

> Sequence inputs (FASTA or index) should be used to filter metadata regardless of the requested outputs.

The current code does not actually fulfill this expectation because it only filters by FASTA sequences when an index is provided or if output sequences are requested. Additionally, the use case above with `--exclude-all` is an exception to this rule where we want to skip all filters and just output the request strains. Maybe the right solution here is to check for `args.sequences` and _not_ `args.exclude_all` before building an index?

Alternately, if we streamed the metadata to disk in the first pass, we could attempt to output all matching sequences for those metadata and then filter metadata on the available sequences when writing metadata output. This approach skips building a sequence index when it isn't needed and only makes a single pass through the sequence data to get strain names.

Reply:
I just tried to address these issues in c256131f93e94eabb91187e33eda2c5736c2bcef and 25f3e9a7036235c554fd869015db12905d672ab1, but I'm curious about feedback from @tsibley, @jameshadfield, and/or @emmahodcroft about your expectations for filter's behavior with respect to filtering by sequence inputs.

Reply:
Those two changes seem reasonable and well-considered to me.